### ISSUE TITLE
fill diagonal in graph

### ISSUE THREAD

**Issue by knaaptime** [pysal/libpysal#668]
is still not yet implemented. The laziest way i can do this now is 

```python
g = Graph.build_kernel(gdf, bandwidth=1000)
sp = g.sparse
sp = sp.tolil()
sp.setdiag(1)
g = Graph.from_sparse(sp, ids=gdf.index)
```

which is what we do currently for W. Not sure if there's something more performant than roundtripping through sparse, given the new format? maybe some pandas-fu or something?

**Comment by sjsrey** [comment]
I'm wondering if we might reconsider the naming convention here? `fill_diagonal` implies a matrix (to me at least), and I think the intention here is to handle the weights on [self-loops](https://networkx.org/documentation/stable/auto_examples/drawing/plot_selfloops.html). 
Given we are all in the Graph, would it be more consistent to call it `fill_self_loops`?

**Comment by sjsrey** [comment]
https://github.com/pysal/libpysal/issues/668#issuecomment-1868027857

**Comment by martinfleis** [comment]
Or `set_self_loops`?

**Comment by knaaptime** [comment]
> I'm wondering if we might reconsider the naming convention here? fill_diagonal implies a matrix (to me at least), and I think the intention here is to handle the weights on [self-loops](https://networkx.org/documentation/stable/auto_examples/drawing/plot_selfloops.html).

I was thinking fill diagonal in matrix terms as well, i.e. set the self-weight to 1 for no discount when multiplying through a kernel graph (so literally filling the diagonal with the self-weight for i,i)

**Comment by knaaptime** [comment]
([just like the one](https://github.com/pysal/libpysal/blob/e24adfbd95507fd1186ab2c8dc7f1feebb870cb9/libpysal/weights/util.py#L837) we have already)

**Comment by knaaptime** [comment]
maybe im misunderstanding though, and the intention is to get away from matrix terminology altogether? in that case, maybe i'd prefer something like `set_self_weight`?

**Comment by sjsrey** [comment]
"Martin Fleischmann" ***@***.***> writes:

> Or set_self_loops?

+1

-- 
Sergio (Serge) Rey
Founding Director, Center for Open Geographical Science
http://cogs.sdsu.edu
Professor of Geography
San Diego State University

**Comment by martinfleis** [comment]
I like @knaaptime's suggestion `set_self_weights` more than my own. It is nicely explicit without a need to think what a self loop is. Self weight is self-explanatory.

**Comment by sjsrey** [comment]
My early comment reflects ptsd around handling the diagonal in the original weights :->

In the case of the new Graph class, the proposed method is returning a new instance of a Graph that includes self-loops with specified weights.

`set_self_weights` implies there are already  self-loops in the Graph and now you are setting  weights on those edges and returning a new Graph.

We already refer to [self-loops](https://github.com/pysal/libpysal/blob/a1ba41dc21b53db1782b5165c1db0f32732b70c9/libpysal/graph/base.py#L990) in Graph when it comes to `isolates`.

Maybe an alternative name for this method  is `add_self_loops` rather than `set_self_loops`? The latter (to me) implies we already have self-loops but we want to change the edge weights. Also, `set` sounds like a setter on the instance which is not what is intended  - we are returning a new instance here not changing the calling instance. Whereas `add_self_loops`  indicates that we want to add self_loops to a new Graph obtained from the current graph.

To summarize - two issues
- I think with Graph we should lean into graph terminology as much as possible
- `set` implying a setter might be confusing?

**Comment by martinfleis** [comment]
@sjsrey that sounds reasonable to me. #670 updated accordingly

**Comment by knaaptime** [comment]
all of that makes sense, but im still not in love with any of the terminology yet. Maybe your graph already has value for self-weights and you just want to change it--or nullify it, in which case you'd be setting self weights to 0 remove the self loop. I dont think `add` is the right verb for the same reason `set` doesnt quite fit

(i also think that, while we want to lean into graph, at some level we also recognize that there's an implicit, ordered sparse matrix representation of the graph, which is why we have ordered IDs. The Graph sits between a networkx network which isnt mathy enough and a scipy sparse, which isnt flexible enough, and sometimes its fair to surface that compromise)

**Comment by knaaptime** [comment]
Between the two, I'd still prefer 'set', because it doesn't necessarily imply an in-place transform to me

**Comment by ljwolf** [comment]
What about "assign_loop_edge_weight()"?

**Comment by martinfleis** [comment]
`assign_self_weight`? Loop_weight does not immediately tell me what it does.

**Comment by knaaptime** [comment]
i think i'd vote for martin's `assign_self_weight` so far. The `weight` is already a standard term on the graph, and the only thing that makes it a 'loop', in graph context is being a self-weight. Since it's the value being assigned not just the cardinality, i prefer that we're setting the *weight*, not the *loop*

**Comment by jGaboardi** [comment]
> i think i'd vote for martin's `assign_self_weight`

If we need a tie breaker, I think @knaaptime sums up the situation well. I'd also vote for:
1. `assign_self_weight`
2. `assign_self_edge_weight` if verbosity is not an issue

**Comment by martinfleis** [comment]
I don't like verbose method names if we can use shorter option. So my vote goes to the shorter one.

**Comment by knaaptime** [comment]
i think we're settled on `assign_self_weight` then

**Comment by jGaboardi** [comment]
> i think we're settled on `assign_self_weight` then

works for me
